### PR TITLE
Add new annotation: @psalm-self-out

### PR DESCRIPTION
**Pull request by olleharstedt**
This PR adds the annotation `@param-self-out` for methods, changing the type of `$this`.

Use-case is to be able to create your own typestates besides the ones already implicit in Psalm (`resource`, and `closed-resource`) in an object-oriented manner.

Typestates are mostly (only?) useful together with alias tracking or ownership inference, which is another discussion: https://github.com/vimeo/psalm/issues/3614

### CODE REVIEW

**Comment by muglug** [comment]
This is a good start. I’d like Psalm to also be able to verify the integrity of `@psalm-self-out` just as it does for `@param-out`.

**Comment by olleharstedt** [comment]
> This is a good start. I’d like Psalm to also be able to verify the integrity of `@psalm-self-out` just as it does for `@param-out`.

Agree.

File: tests/SelfOutTest.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+                  interface Foo {
+                      /**
+                       * @return void
+                       */
+                      public function far() {
+                      }
+                  }
+                  class Bar {
+                      /**
+                       * @psalm-self-out Foo
```
[the remark this answers is not in the document]
Comment (olleharstedt) [reply tests/SelfOutTest.php:25]: Yes, but maybe let's detail a little bit what we mean? Using the closed/open file as example:

```php
interface ClosedFile {
  /**
   * @psalm-self-out File&OpenFile
   */
  public function open() {}
}
interface OpenFile {
  /**
   * @psalm-self-out File&ClosedFile
   */
  public function close() {}
  public function read() {}
}
class File implements OpenFile, ClosedFile {
  ... function defs here
}
```

Then we define `psalm-self-out` as always requiring first part of the intersection type being the base-class, second part being the "next" state. This way, Psalm doesn't need to "remember" the base-class in a separate field.

[the remark this answers is not in the document]
Comment (weirdan) [reply tests/SelfOutTest.php:25]: But in this case `File` is already `File&OpenFile&ClosedFile`, by virtue of implementing those interfaces, isn't it? Or do you propose to reuse the same syntax to achieve different effects?

[the remark this answers is not in the document]
Comment (olleharstedt) [reply tests/SelfOutTest.php:25]: Oh, you're right. Hm, so maybe intersection types cannot be used.

> Or do you propose to reuse the same syntax to achieve different effects?

No no no.

**Comment by olleharstedt** [comment]
Fun fact: In Smalltalk something similar is called `become`: https://gbracha.blogspot.com/2009/07/miracle-of-become.html

**Comment by olleharstedt** [comment]
@muglug Can you help me with this error message? What's the difference between type and storage type? Can one be cast to the other?

> ERROR: InvalidPropertyAssignmentValue - src/Psalm/Internal/Analyzer/Statements/Expression/Call/MethodCallAnalyzer.php:415:21 - $context->vars_in_scope with declared type 'array<string, Psalm\Type\Union>' cannot be assigned type 'non-empty-array<string, Psalm\Storage\Type\Union|Psalm\Type\Union>' (see https://psalm.dev/145)

**Comment by olleharstedt** [comment]
@muglug Hm, how can I debug Circle CI without giving it access to absolutely everything on my account?

**Comment by olleharstedt** [comment]
Only thing left failing is taint tests, it seems.

**Comment by muglug** [comment]
> Only thing left failing is taint tests, it seems.

Ignore that, but would you mind squashing these commits?

**Comment by olleharstedt** [comment]
> > Only thing left failing is taint tests, it seems.
> 
> Ignore that, but would you mind squashing these commits?

Can't do it without rebasing everthing. Can't you squash and merge instead if/when this PR gets accepted?

**Comment by muglug** [comment]
Thanks!

**Pull request merged**